# Проект по SQL

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Компания решила быть на волне и купила крупный сервис для чтения книг по подписке.

**Цель исследования** - проанализировать базу данных, ответить на поставленные вопросы, сделать выводы для дальнейшей формулироваки ценностного предложения для нового продукта.

In [1]:
import pandas as pd
from sqlalchemy import create_engine
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'], db_config['pwd'], db_config['host'], db_config['port'],
db_config['db']) 
engine = create_engine(connection_string, connect_args={'sslmode':'require'})


In [2]:
data = '''
select *
from books
limit 10;'''


In [3]:
books= pd.io.sql.read_sql(data, con = engine)
books

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   
5        6        257               1st to Die (Women's Murder Club  #1)   
6        7        258               2nd Chance (Women's Murder Club  #2)   
7        8        260              4th of July (Women's Murder Club  #4)   
8        9        563                                   A Beautiful Mind   
9       10        445                                 A Bend in the Road   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268  
5        424       2005-05-20           116  
6        400       2005-05-20           116  
7        448       2006-06-01           318  
8        461       2002-02-04           104  
9        341       2005-04-01           116

In [4]:
data = '''
select *
from authors
limit 10;'''

In [5]:
authors= pd.io.sql.read_sql(data, con = engine)
authors

author_id                                author
0          1                            A.S. Byatt
1          2        Aesop/Laura Harris/Laura Gibbs
2          3                       Agatha Christie
3          4                         Alan Brennert
4          5              Alan Moore/David   Lloyd
5          6                            Alan Paton
6          7           Albert Camus/Justin O'Brien
7          8                         Aldous Huxley
8          9    Aldous Huxley/Christopher Hitchens
9         10  Aleksandr Solzhenitsyn/H.T. Willetts

In [6]:
data = '''
select *
from publishers
limit 10;'''

In [7]:
publishers= pd.io.sql.read_sql(data, con = engine)
publishers

publisher_id                                publisher
0             1                                      Ace
1             2                                 Ace Book
2             3                                Ace Books
3             4                            Ace Hardcover
4             5        Addison Wesley Publishing Company
5             6                                  Aladdin
6             7                       Aladdin Paperbacks
7             8                             Albin Michel
8             9                          Alfred A. Knopf
9            10  Alfred A. Knopf Books for Young Readers

In [8]:
data = '''
select *
from ratings
limit 10;'''

In [9]:
ratings= pd.io.sql.read_sql(data, con = engine)
ratings

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2
5          6        3  johnsonamanda       4
6          7        3    scotttamara       5
7          8        3    lesliegibbs       5
8          9        4    abbottjames       5
9         10        4   valenciaanne       4

In [10]:
data = '''
select *
from reviews
limit 10;'''

In [11]:
reviews= pd.io.sql.read_sql(data, con = engine)
reviews

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   
5          6        3    lesliegibbs   
6          7        4   valenciaanne   
7          8        4    abbottjames   
8          9        5        npowers   
9         10        5        staylor   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...  
5           Analysis no several cause international.  
6  One there cost another. Say type save. With pe...  
7  Within enough mother. There at system full rec...  
8  Thank now focus realize economy focus fly. Ite...  
9  Game push lot reduce where remember. Including...

На данном этапе выгрузили данные, изучили содержимое таблиц из базы данных, переходим к анализу данных и ответам на поставленные вопросы


### Необходимо посчитать, сколько книг вышло после 1 января 2000 года

In [12]:
data = '''
select count (book_id)
from books
where publication_date > '2000-01-01'
'''

In [13]:
books_count= pd.io.sql.read_sql(data, con = engine)
books_count

count
0    819

После 1 января 2000 года было выпущено 819 книг

### Необходимо посчитать количество обзоров и среднюю оценку для каждой книги

In [28]:
data = '''
select b.title,count (distinct r.review_id), round(avg(ra.rating), 2) as rating
from reviews as r
right outer join books as b on r.book_id=b.book_id
left outer join ratings as ra on b.book_id = ra.book_id
group by b.book_id
order by rating desc

'''

In [29]:
books_rat= pd.io.sql.read_sql(data, con = engine)
books_rat

title  count  rating
0        Arrows of the Queen (Heralds of Valdemar  #1)      2    5.00
1    The Walking Dead  Book One (The Walking Dead #...      2    5.00
2                                      Light in August      2    5.00
3    Wherever You Go  There You Are: Mindfulness Me...      2    5.00
4    Captivating: Unveiling the Mystery of a Woman'...      2    5.00
..                                                 ...    ...     ...
995  The World Is Flat: A Brief History of the Twen...      3    2.25
996                  His Excellency: George Washington      2    2.00
997                                      Drowning Ruth      3    2.00
998                                              Junky      2    2.00
999                               Harvesting the Heart      2    1.50

[1000 rows x 3 columns]

Было обозначено ТОП-10 книг по рейтингу. В данных довольно много книг с рейтингом 5.0, но у таких книг 2 обзора.
При этом, если обозначить ТОП книг по кол-ву отзывов, то наибольшее количество обзоров у книги Twilight. 
Среди прочего выделяется книга A Dirty Job у которой 4 обзора и рейтинг 5.0. Наименьшим рейтингом обладает книга Harvesting the Heart (рейтинг - 1.5)

### Необходимо определить издательство, которое выпустило наибольшее число книг толще 50 страниц

In [16]:
data = '''
select p.publisher, count(b.book_id) as books_num
from publishers as p
join books as b on b.publisher_id = p.publisher_id
where b.num_pages > 50
group by p.publisher
order by books_num desc
limit 1
'''

In [17]:
pub= pd.io.sql.read_sql(data, con = engine)
pub

publisher  books_num
0  Penguin Books         42

Издательство Penguin Books выпустило наибольшее количество книг толщиной более 50 страниц

### Необходимо определить автора с самой высокой средней оценкой книг (с 50+ оценками)

In [18]:
data = '''
select a.author, avg(r.rating) as rating
from authors as a
join books as b on b.author_id = a.author_id
join ratings as r on b.book_id = r.book_id
where b.book_id in
    (select b.book_id
     from books AS b
     join ratings AS r ON b.book_id=r.book_id
     group by b.book_id
     having count(r.rating_id)>=50)
group by a.author
order by rating desc

limit 1
'''

In [19]:
author= pd.io.sql.read_sql(data, con = engine)
author

author    rating
0  J.K. Rowling/Mary GrandPré  4.287097

J.K. Rowling/Mary GrandPré является автором с самой высокой средней оценкой книг.

### Необходимо посчитайть среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [20]:
data = '''
with t1 as(
select username, count(review_id) as reviews
from reviews
where username in (
select username
        from ratings
        group by username
        having count(rating_id)>50)
group by username
)
SELECT round (SUM (reviews) / COUNT(reviews), 0) AS AVG
from t1
'''

In [21]:
rat= pd.io.sql.read_sql(data, con = engine)
rat

avg
0  24.0

Среднее количество обзоров от пользователей, которые поставили больше 50 оценок составляет 24.

# Выводы

Таким образом, было проведено комплексное исследование, были даны ответы на поставленные заказчиком исследования вопросы.